core/templates/services/messenger.service.ts

```diff
@@ -181,16 +194,56 @@ export class MessengerService {
 
       if (this.SUPPORTED_HASHDICT_VERSIONS.has(hashDict.version)) {
         this.loggerService.info('Posting message to parent: ' + messageTitle);
+        let payload: PayloadType;
+        let isValidMessage: boolean;
+        switch (
+          <keyof typeof ServicesConstants.MESSENGER_PAYLOAD> messageTitle) {
+          case 'EXPLORATION_COMPLETED':
+            payload = this.getPayload.explorationCompleted(
+              <ExplorationCompletedData> messageData);
+            isValidMessage = (
+              this.MESSAGE_VALIDATORS.explorationCompleted());
+            break;
+          case 'EXPLORATION_LOADED':
+            payload = this.getPayload.explorationLoaded(
+              <ExplorationLoadedData> messageData);
+            isValidMessage = (
+              this.MESSAGE_VALIDATORS.explorationLoaded());
+            break;
+          case 'EXPLORATION_RESET':
+            payload = this.getPayload.explorationReset(
+              <string> messageData);
+            isValidMessage = (
+              this.MESSAGE_VALIDATORS.explorationReset(payload));
+            break;
+          case 'HEIGHT_CHANGE':
+            payload = this.getPayload.heightChange(
+              <HeightChangeData> messageData);
+            isValidMessage = (
+              this.MESSAGE_VALIDATORS.heightChange(payload));
+            break;
+          case 'STATE_TRANSITION':
+            payload = this.getPayload.stateTransition(
+              <StateTransitionData> messageData);
+            isValidMessage = (
+              this.MESSAGE_VALIDATORS.stateTransition(
+                <StateTransitionData> payload));
+            break;
+        }
```

Comment:
I could not come up with a simpler approach since the strict checks don't pass due to the fact that each method takes a different type of data as an argument and that type explicitly needs to be separated.

if we use it as a conjunction;
```
let payload = this.getPayload[messageTitle](messageData);
``` 
then the compiler expects `messageData`  to be of type `HeightChangeData & ExplorationCompletedData & StateTransitionData & so on` (since the method to use will be decided on runtime) and not `HeightChangeData | ExplorationCompletedData | StateTransitionData | so on`

Reply:
Ah, okay, that makes sense.